src/main/java/com/epam/esm/service/CertificateService.java

```diff
@@ -27,78 +28,73 @@
 public class CertificateService {
     private final CertificateRepository certificateRepository;
     private final TagRepository tagRepository;
+    private final SpringDataTagRepository springDataTagRepository;
     private final CertificateDTOAssembler certificateDTOMapper;
+    private final SpringDataGiftCertificateRepository giftCertificateRepository;
+    private final EntityManager entityManager;
+
+    public GiftCertificate getCertificate(long id) {
+        return giftCertificateRepository.findById(id).orElseThrow(() -> new EntityNotFoundException("can't receive GiftCertificate with id = " + id, Error.GiftCertificateNotFound));
 
-    public GiftCertificateDTO getCertificate(long id) {
-        GiftCertificate certificate = certificateRepository.fetchCertificate(id).orElseThrow(() -> new EntityNotFoundException("can't receive GiftCertificate with id = " + id, Error.GiftCertificateNotFound));
-        List<Tag> tags = tagRepository.fetchLinkedTags(id);
-        return certificateDTOMapper.mapToDTO(certificate, tags);
     }
 
     @Transactional
-    public GiftCertificateDTO addCertificate(GiftCertificateDTO certificateDTO) {
-        GiftCertificate giftCertificate = certificateDTOMapper.mapToCertificate(certificateDTO);
-        GiftCertificate certificateFromDb = certificateRepository.createCertificate(giftCertificate);
-        tagRepository.saveTagsForCertificate(certificateDTO.getTags(), certificateFromDb.getId());
-        certificateFromDb = certificateRepository.fetchCertificate(certificateFromDb.getId()).orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST));
-        List<Tag> tagsFromDb = tagRepository.fetchLinkedTags(certificateFromDb.getId());
-        return certificateDTOMapper.mapToDTO(certificateFromDb, tagsFromDb);
+    public GiftCertificate addCertificate(GiftCertificate certificate) {
+        // TODO Is it possible to do it in one row? Without usage of tagsRepo or EntityManager??
+        List<Tag> tags = certificate.getTags();
+        tags = springDataTagRepository.saveAll(tags);
+        tags.stream()
+                .forEach(entityManager::refresh);
+        certificate.setTags(tags);
```

Comment:
Hm you can try to collapse tags oparation to one stream flow

Reply:
But it's not necessary

Reply:
ok, thanks. Understood